lib/auth/client_tls_config_generator.go

```diff
@@ -0,0 +1,272 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"log/slog"
+	"math"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/genmap"
+	"github.com/gravitational/trace"
+)
+
+// ClientTLSConfigGeneratorConfig holds parameters for ClientTLSConfigGenerator setup.
+type ClientTLSConfigGeneratorConfig struct {
+	// TLS is the upstream TLS config that per-cluster configs are generated from.
+	TLS *tls.Config
+	// ClusterName is the name of the current cluster.
+	ClusterName string
+	// PermiteRemoteClusters must be true for non-local cluster CAs to be used. Most usecases
+	// want this to be true, but it defaults to false to help avoid accidentally expanding the
+	// CA pool in cases where remote cluster CA usage is inappropriate.
+	PermitRemoteClusters bool
+	// AccessPoint is the upstream data source used to lookup cert authorities
+	// and watch for changes.
+	AccessPoint AccessCacheWithEvents
+	// Context is an optional parent context that will be used when setting up the internal
+	// close context of the generator if supplied.
+	Context context.Context
```

Comment:
Can we not stash contexts in structs :pray:?

I get that its convenient and we do it all over the place, but if we don't start writing new code any differently the story will never improve.

You don't need most of what context provides. The advice here seems to fit your situation too: https://groups.google.com/g/golang-nuts/c/xRbzq8yzKWI/m/gnELT-YQXmQJ